src/vm_pmu.rs

```diff
@@ -365,7 +368,112 @@ impl VmPmuState {
+        config_flags: PmuCounterConfigFlags,
+        event_type: PmuEventType,
+        event_data: u64,
+    ) -> SbiResult<u64> {
+        use sbi::PmuHardware::*;
+        fn configure_counter_csrs(
+            config_flags: PmuCounterConfigFlags,
+            event_type: PmuEventType,
+            event_data: u64,
+            counter_index: u64,
+        ) {
+            let mut shpmevent = LocalRegisterCopy::<u64, shpmevent::Register>::new(0);
+            // Translate uinh/sinh
+            if config_flags.is_uinh() || config_flags.is_vuinh() {
+                shpmevent.modify(shpmevent::vuinh.val(1));
+            }
+            if config_flags.is_sinh() || config_flags.is_vsinh() {
+                shpmevent.modify(shpmevent::vsinh.val(1));
+            }
+            shpmevent.modify(shpmevent::sinh.val(1));
+            // TODO: The mapping of events to platform counters is still under discussion.
+            // This will need to be revisited after the standard has been finalized.
+            match event_type {
+                PmuEventType::Hardware(_) | PmuEventType::Cache(_) => {
+                    shpmevent.modify(shpmevent::event_selector.val(event_type.raw()))
+                }
+                PmuEventType::RawEvent => {
+                    shpmevent.modify(shpmevent::event_selector.val(event_data))
+                }
+                _ => unreachable!(),
+            }
+            CSR.scounterselect.set(counter_index);
+            CSR.shpmevent.set(shpmevent.get());
+            if config_flags.is_clear_value() {
+                CSR.shpmcounter.set(0);
+            }
+            if config_flags.is_auto_start() {
+                CSR.scountinhibit.read_and_clear_bits(1 << counter_index);
+            }
+        }
+
+        let pmu_info = pmu::PmuInfo::get()?;
+        let counter_mask = pmu_info.filter_counter_mask(counter_index, counter_mask)?;
+        let index = if !config_flags.is_skip_match() {
+            match event_type {
+                PmuEventType::Hardware(hardware_event_type)
+                    if matches!(hardware_event_type, CpuCycles)
+                        || matches!(hardware_event_type, Instructions) =>
+                {
+                    let mut iter = CounterMaskIter::new(counter_index, counter_mask);
+                    // Fixed counters must use a static mapping pending finalization of event mapping
+                    let platform_index = if matches!(hardware_event_type, CpuCycles) {
+                        const CYCLE: u64 = 0;
+                        CYCLE
+                    } else {
+                        const INSRET: u64 = 2;
+                        INSRET
+                    };
+                    if iter.any(|index| {
+                        index == platform_index as usize
+                            && matches!(self.counter_state[index], PmuCounterState::NotConfigured)
+                            && pmu_info.get_counter_info(platform_index).is_ok()
```

Comment:
May be I was not clear in early attempts. Without delegation, any counter can not be accessed/updated. So that's always my implicit assumption. As said in the above comment, 
Statement from the counter delegation spec
"When HPMDE=1, smcounteren is shadow of the mcounteren register, otherwise reads return 0."

and priv spec
"When the CY, TM, IR, or HPMn bit in the mcounteren register is clear, attempts to read the
cycle, time, instret, or hpmcountern register while executing in S-mode or U-mode will cause
an illegal instruction exception. When one of these bits is set, access to the corresponding register
is permitted in the next implemented privilege mode (S-mode if implemented, otherwise U-mode)."

Thus, we shouldn't skip the delegation check at any point. 

TLDR; we should pick option "a" pointed by @abrestic-rivos

Reply:
FWIW, we were enforcing the CYCLE/INSRET delegation check before if was removed because it wasn't outlined in the specification. We should update it in the spec as well.